exercises/practice/circular-buffer/.meta/proof.ci.wat

```diff
@@ -1,31 +1,59 @@
 (module
-  (memory 1)
+  ;; a WebAssembly page is 64KiB, so each page holds up to 16384 i32s
+  ;; Our linear memory is one page by default, but it is permitted to grow
+  ;; up to four pages via use of the memory.grow instruction, which can hold
+  ;; up to 65536 i32s.
+  (memory (export "mem") 1 4)
   (global $head (mut i32) (i32.const -1))
   (global $tail (mut i32) (i32.const -1))
   (global $capacity (mut i32) (i32.const 0))
   (global $i32Size i32 (i32.const 4))
 
-
-  ;; capacity: the number of elements to store
-  ;; elementSize: the size of the element to store in bytes
-  ;; Does not support resizing circular buffer. Wipes all data
+  ;;
+  ;; Initialize a circular buffer of i32s with a given capacity
+  ;;
+  ;; @param {i32} newCapacity - capacity of the circular buffer between 0 and 65,536
+  ;;                            in order to fit in four 64KiB WebAssembly pages.
+  ;;
+  ;; @returns {i32} 0 on success or -1 on error
+  ;; 
   (func (export "init") (param $newCapacity i32) (result i32)
-    ;; a WebAssembly page is 4096 bytes, so up to 1024 i32s
-    (if (i32.gt_s (local.get $newCapacity) (i32.const 1024)) (then
-      (return (i32.const -1))
-    ))
+    ;; a WebAssembly page is 64KiB, so each page holds up to 16384 i32s
+    ;; Our linear memory can grow up to four pages, so we can hold up to 65536 i32s
+    (if (i32.or 
+      (i32.lt_s (local.get $newCapacity) (i32.const 0)) 
+      (i32.gt_s (local.get $newCapacity) (i32.const 65536))) (then
+        (return (i32.const -1))))
 
     (global.set $head (i32.const -1))
     (global.set $tail (i32.const -1))
     (global.set $capacity (local.get $newCapacity))
-    (i32.const 0)
+
+    ;; We do not need to grow the memory if the new capacity is less than 16384
+    (if (result i32) (i32.le_s (local.get $newCapacity) (i32.const 16384)) (then
+      (i32.const 0)
+    ) (else 
+      ;; memory.grow returns old size on success or -1 on failure
+      (memory.grow (i32.div_s (i32.sub (local.get $newCapacity) (i32.const 1)) (i32.const 16384)))
```

Comment:
Do you want `(newCap - 1) / 16384` or `(newCap / 16384) - 1` here? I think the latter.

Reply:
The latter actually ends up allocating too few pages I believe due to integer division acting like a floor:

x idiv y ~= floor(x/y)

i.e. when we want to allocate a circular buffer that holds one more than we can fit in 1 page.

(newCap - 1) / 16384
((16385 - 1) idiv 16384) => (16384 idiv 16384) => 1

(newCap / 16384) - 1
((16385 idiv 16384) - 1) => (1 - 1) => 0

I checked this by modifying the solution to `(memory.grow (i32.sub (i32.div_s (local.get $newCapacity) (i32.const 16384)) (i32.const 1)))` and got the following:

```
    ✓ initializing negative capacity should error (2 ms)
    ✓ initializing capacity of 0 i32s should result in a linear memory with 1 page (1 ms)
    ✓ initializing capacity of 16384 i32s should result in a linear memory with 1 page
    ✕ initializing capacity of 16385 i32s should result in a linear memory with 2 pages (1 ms)
    ✓ initializing capacity of 32768 i32s should result in a linear memory with 2 pages (1 ms)
    ✕ initializing capacity of 32769 i32s should result in a linear memory with 3 pages
    ✓ initializing capacity of 49152 i32s should result in a linear memory with 3 pages
    ✕ initializing capacity of 49153 i32s should result in a linear memory with 4 pages
    ✓ initializing capacity of 65536 i32s should result in a linear memory with 4 pages
    
      ● CircularBuffer › initializing capacity of 16385 i32s should result in a linear memory with 2 pages

    expect(received).toEqual(expected) // deep equality

    Expected: 2
    Received: 1

      47 |   test("initializing capacity of 16385 i32s should result in a linear memory with 2 pages", () => {
      48 |     expect(currentInstance.exports.init(16385)).toEqual(0);
    > 49 |     expect(currentInstance.exports.mem.buffer.byteLength / 65536).toEqual(2);
         |                                                                   ^
      50 |   });
      51 |
      52 |   test("initializing capacity of 32768 i32s should result in a linear memory with 2 pages", () => {

      at Object.toEqual (circular-buffer.spec.js:49:67)

  ● CircularBuffer › initializing capacity of 32769 i32s should result in a linear memory with 3 pages

    expect(received).toEqual(expected) // deep equality

    Expected: 3
    Received: 2

      57 |   test("initializing capacity of 32769 i32s should result in a linear memory with 3 pages", () => {
      58 |     expect(currentInstance.exports.init(32769)).toEqual(0);
    > 59 |     expect(currentInstance.exports.mem.buffer.byteLength / 65536).toEqual(3);
         |                                                                   ^
      60 |   });
      61 |
      62 |   test("initializing capacity of 49152 i32s should result in a linear memory with 3 pages", () => {

      at Object.toEqual (circular-buffer.spec.js:59:67)

  ● CircularBuffer › initializing capacity of 49153 i32s should result in a linear memory with 4 pages

    expect(received).toEqual(expected) // deep equality

    Expected: 4
    Received: 3

      67 |   test("initializing capacity of 49153 i32s should result in a linear memory with 4 pages", () => {
      68 |     expect(currentInstance.exports.init(49153)).toEqual(0);
    > 69 |     expect(currentInstance.exports.mem.buffer.byteLength / 65536).toEqual(4);
         |                                                                   ^
      70 |   });
      71 |
      72 |   test("initializing capacity of 65536 i32s should result in a linear memory with 4 pages", () => {

      at Object.toEqual (circular-buffer.spec.js:69:67)
```